src/ProductConstructionService/ProductConstructionService.FeedCleaner/FeedCleanerJob.cs

```diff
@@ -0,0 +1,161 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Text.RegularExpressions;
+using Kusto.Cloud.Platform.Utils;
+using Microsoft.DotNet.DarcLib;
+using Microsoft.DotNet.DarcLib.Helpers;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+using PackagesInReleaseFeeds = System.Collections.Generic.Dictionary<string, System.Collections.Generic.Dictionary<string, System.Collections.Generic.HashSet<string>>>;
+
+namespace ProductConstructionService.FeedCleaner;
+
+public class FeedCleanerJob
+{
+    private readonly IServiceProvider _serviceProvider;
+    private readonly IAzureDevOpsClient _azureDevOpsClient;
+    private readonly IOptions<FeedCleanerOptions> _options;
+    private readonly Microsoft.Extensions.Logging.ILogger<FeedCleanerJob> _logger;
+
+    public FeedCleanerJob(
+        IServiceProvider serviceProvider,
+        IAzureDevOpsClient azureDevOpsClient,
+        IOptions<FeedCleanerOptions> options,
+        Microsoft.Extensions.Logging.ILogger<FeedCleanerJob> logger)
+    {
+        _serviceProvider = serviceProvider;
+        _azureDevOpsClient = azureDevOpsClient;
+        _options = options;
+        _logger = logger;
+    }
+
+    private FeedCleanerOptions Options => _options.Value;
+
+    public async Task CleanManagedFeedsAsync()
+    {
+        if (!Options.Enabled)
+        {
+            _logger.LogInformation("Feed cleaner service is disabled in this environment");
+            return;
+        }
+
+        _logger.LogInformation("Loading packages in release feeds...");
+
+        PackagesInReleaseFeeds packagesInReleaseFeeds = await GetPackagesForReleaseFeedsAsync();
+
+        _logger.LogInformation("Loaded {versionCount} versions of {packageCount} packages from {feedCount} feeds",
+            packagesInReleaseFeeds.Sum(feed => feed.Value.Sum(package => package.Value.Count)),
+            packagesInReleaseFeeds.Sum(feed => feed.Value.Keys.Count),
+            packagesInReleaseFeeds.Keys.Count);
+
+        foreach (var azdoAccount in Options.AzdoAccounts)
+        {
+            _logger.LogInformation("Processing feeds for {account}...", azdoAccount);
+
+            List<AzureDevOpsFeed> allFeeds;
+            try
+            {
+                allFeeds = await _azureDevOpsClient.GetFeedsAsync(azdoAccount);
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError(ex, "Failed to get feeds for account {azdoAccount}", azdoAccount);
+                continue;
+            }
+
+            List<AzureDevOpsFeed> managedFeeds = allFeeds
+                .Where(f => Regex.IsMatch(f.Name, FeedConstants.MaestroManagedFeedNamePattern))
+                .Shuffle()
```

Comment:
I didn't quite understand why we need to shuffle the feeds we get from AzDO. Could you please expand on this?

Reply:
Before the other optimizations, the feed cleaner would not get to all the feeds in time and it would keep processing the same ones over and over (the packages in them could not have been removed so it kept going over those). I added the shuffle so that all feeds get looked at. But with the parallelism it can now get to all, however I think it's still maybe a good idea to shuffle.

Reply:
Thanks 🙇